packages/hash/api/src/model/knowledge/linkEntity.model.ts

```diff
@@ -0,0 +1,187 @@
+import {
+  GraphApi,
+  Entity,
+  LinkEntityMetadata,
+  Filter,
+  EntityStructuralQuery,
+} from "@hashintel/hash-graph-client";
+
+import { EntityModel, EntityTypeModel, LinkEntityModel } from "../index";
+
+export type LinkModelConstructorParams = {
+  linkEntity: Entity;
+  linkEntityTypeModel: EntityTypeModel;
+  leftEntityModel: EntityModel;
+  rightEntityModel: EntityModel;
+};
+
+export type LinkModelCreateParams = {
+  ownedById: string;
+  linkEntityTypeModel: EntityTypeModel;
+  leftEntityModel: EntityModel;
+  leftOrder?: number;
+  rightEntityModel: EntityModel;
+  rightOrder?: number;
+  actorId: string;
+};
+
+/**
+ * @class {@link LinkEntityModel}
+ */
+export default class extends EntityModel {
+  leftEntityModel: EntityModel;
+
+  rightEntityModel: EntityModel;
+
+  get linkMetadata(): LinkEntityMetadata {
+    return this.entity.metadata.linkMetadata!;
+  }
+
+  constructor({
+    linkEntity,
+    linkEntityTypeModel,
+    leftEntityModel,
+    rightEntityModel,
+  }: LinkModelConstructorParams) {
+    super({ entity: linkEntity, entityTypeModel: linkEntityTypeModel });
+    this.leftEntityModel = leftEntityModel;
+    this.rightEntityModel = rightEntityModel;
+  }
+
+  static async fromEntity(
```

Comment:
I get the sense that this is how we're going to use this model class most of the time. But ideally we'd have a structural query with appropriate resolve depths that mean we don't need to do all of the `Promise.all` below because we'd already have it.

This is probably too big a change to do in this PR as a drive-by, but thought I'd highlight it.

We might want to change a lot of the model class impls to actually take in a subgraph and only make follow-up requests if they can't find it in the subgraph.

Reply:
This is a reminder to myself, I should paste a link to this comment in the asana task about revisiting model classes. (Can't do it immediately because the comment is pending so it doesn't have a link 🙃)

Reply:
I have added a link back to this comment in the asana task for the model classes ergonomics thing.

Reply:
Yes, there's definitely room for optimization here - but think we should only address after this model class discussion as it's an investment.